### PR TITLE
fix(HACBS-2549): operation EnsureReleaseIsValid is buggy

### PR DESCRIPTION
**Pull request by johnbieren**
The EnsureReleaseIsValid operation uses nested patches. This leads to some unexpected behavior. This commit breaks the operation into multiple validation operations to work around the bug.

### CODE REVIEW

File: controllers/release/adapter.go [hunk 1]
```diff
@@ -69,6 +69,39 @@ func newAdapter(ctx context.Context, client client.Client, release *v1alpha1.Rel
...
+		if errors.IsNotFound(err) {
+			a.release.MarkValidationFailed(err.Error())
+			a.release.MarkReleaseFailed("Release validation failed")
+			return controller.RequeueOnErrorOrStop(a.client.Status().Patch(a.ctx, a.release, patch))
+		}
+
+		return controller.RequeueWithError(err)
+	}
+
+	err = a.registerAttributionData(releasePlan)
```
Comment (davidmogar) [inline controllers/release/adapter.go:95]: Once again you have a patch within a patch :stuck_out_tongue_winking_eye:

File: controllers/release/adapter.go [hunk 2]
```diff
@@ -69,6 +69,39 @@ func newAdapter(ctx context.Context, client client.Client, release *v1alpha1.Rel
...
+	}
+
+	if a.release.Labels[metadata.AutomatedLabel] == "true" && !a.release.IsAutomated() {
+		return controller.RequeueWithError(fmt.Errorf("automated not set in status for automated release"))
+	}
+
+	patch := client.MergeFrom(a.release.DeepCopy())
+	releasePlan, err := a.loader.GetReleasePlan(a.ctx, a.client, a.release)
+	if err != nil {
+		if errors.IsNotFound(err) {
```
Comment (davidmogar) [inline controllers/release/adapter.go:86]: This error handling could be merged with the one in L96 so we don't duplicate that much code. This whole validation thing is so annoying... I'm really considering to follow the approach of the validate functions with the same signature so we can loop over them. We can pass the resources to all of them and return an error. At the point of validate stuff, the processing resources and the release is everything a validation function could need.

Comment (johnbieren) [reply controllers/release/adapter.go:86]: how? we don't want to call line 95 unless the error is not nil from line 84

File: controllers/release/adapter.go (hunk 1 above)
Comment (johnbieren) [reply controllers/release/adapter.go:95]: fixed. removed patch from registerAttributionData

File: controllers/release/adapter.go (hunk 2 above)
Comment (davidmogar) [reply controllers/release/adapter.go:86]: Changing it completely ^^

File: controllers/release/adapter.go (hunk 1 above)
Comment (davidmogar) [reply controllers/release/adapter.go:95]: As we said offline, `register` functions should actually use patch.

File: controllers/release/adapter.go [hunk 3]
```diff
@@ -491,38 +488,60 @@ func (a *adapter) syncResources() error {
...
+func (a *adapter) validateProcessingResources() (valid bool, err error) {
+	resources, err := a.loader.GetProcessingResources(a.ctx, a.client, a.release)
+	if err != nil {
+		if resources == nil || resources.ReleasePlanAdmission == nil || errors.IsNotFound(err) {
+			a.release.MarkValidationFailed(err.Error())
+			return false, nil
+		}
 
-	return a.registerAttributionData(releasePlan)
+		return false, err
```
Comment (johnbieren) [inline controllers/release/adapter.go:544]: @davidmogar I don't know how to mock the loader to get to this part of the logic, so there is no test for it. Any ideas?

File: controllers/release/adapter.go [hunk 4]
```diff
@@ -246,36 +246,33 @@ func (a *adapter) EnsureReleaseIsProcessed() (controller.OperationResult, error)
...
+		valid, err := validationFunc()
+		if err != nil {
 			return controller.RequeueWithError(err)
 		}
-
-		a.release.MarkValidationFailed(err.Error())
-		a.release.MarkReleaseFailed("Author validation failed")
-		return controller.RequeueOnErrorOrStop(a.client.Status().Patch(a.ctx, a.release, patch))
+		if !valid {
+			a.release.MarkReleaseFailed("Release validation failed")
```
Comment (davidmogar) [inline controllers/release/adapter.go:265]: Nice! I like it here.

File: controllers/release/adapter.go (hunk 3 above)
Comment (davidmogar) [reply controllers/release/adapter.go:544]: I mean, you just need to return an error resources (the releasePlanAdmission is enough probably), isn't it? Something like the following:
```
{
    ContextKey: loader.ProcessingResourcesContextKey,
    Err:        fmt.Errorf("internal error")
    Resource:   &loader.ProcessingResources{
        ReleasePlanAdmission: releasePlanAdmission,
    },
},
```

Comment (johnbieren) [reply controllers/release/adapter.go:544]: thanks, tests added

File: controllers/release/adapter.go (hunk 4 above)
Comment (davidmogar) [inline controllers/release/adapter.go:265]: Should we `break` here? No point on continuing.

File: controllers/release/adapter_test.go [hunk 5]
```diff
@@ -1241,31 +1238,42 @@ var _ = Describe("Release adapter", Ordered, func() {
...
+		It("should return invalid and no error if the ReleasePlan is not found", func() {
+			adapter.ctx = toolkit.GetMockedContext(ctx, []toolkit.MockData{
+				{
+					ContextKey: loader.ReleasePlanContextKey,
+					Err:        errors.NewNotFound(schema.GroupResource{}, ""),
+				},
+			})
+
+			valid, err := adapter.validateAuthor()
+			Expect(!valid).To(BeTrue())
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:1275]: Man, `Expect(valid).To(BeFalse())` please :stuck_out_tongue_winking_eye:

File: controllers/release/adapter_test.go [hunk 6]
```diff
@@ -1275,27 +1283,73 @@ var _ = Describe("Release adapter", Ordered, func() {
 				}
 			})
 
-			It("returns an error if the ReleasePlan has no author", func() {
+			It("returns invalid and an error if automated label is present but is not set in release status", func() {
+				valid, err := adapter.validateAuthor()
+				Expect(!valid).To(BeTrue())
+				Expect(err).To(HaveOccurred())
+				for i := range adapter.release.Status.Conditions {
+					if adapter.release.Status.Conditions[i].Type == "Validated" {
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:1291]: Maybe we should expose the condition types so you can do `FindStatusCondition(adapter.release.Status.Conditions, v1alpha1.ReleaseValidatedConditionType)`. Probably we should create a ticket for the future to see if we want to tackle it.

File: controllers/release/adapter_test.go [hunk 7]
```diff
@@ -656,26 +646,32 @@ var _ = Describe("Release adapter", Ordered, func() {
...
-			Expect(err).NotTo(HaveOccurred())
+			Expect(result.RequeueRequest && !result.CancelRequest).To(BeTrue())
+			Expect(err).To(HaveOccurred())
 			Expect(adapter.release.IsValid()).To(BeFalse())
-			Expect(adapter.release.HasReleaseFinished()).To(BeTrue())
+			Expect(adapter.release.HasReleaseFinished()).To(BeFalse())
 		})
 
-		It("should stop reconcile if multiple ReleasePlanAdmissions exist", func() {
+		It("should stop reconcile if validateProcessingResources marks the release invalid", func() {
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:670]: It sucks that we have to check this here. Maybe a ticket to think about how to remove this from here (and add a proper validator) in the future?

File: controllers/release/adapter_test.go [hunk 8]
```diff
@@ -1275,27 +1283,73 @@ var _ = Describe("Release adapter", Ordered, func() {
...
+				Expect(conditionMsg).To(Equal("automated not set in status for automated release"))
+			})
+
+			It("returns invalid and an error if the ReleasePlan has no author", func() {
 				adapter.release.Status.Automated = true
-				err := adapter.registerAttributionData(releasePlan)
-				Expect(err).NotTo(BeNil())
-				Expect(err.Error()).To(ContainSubstring("no author in the ReleasePlan"))
+				valid, err := adapter.validateAuthor()
+				Expect(!valid).To(BeTrue())
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:1301]:
```suggestion
				Expect(valid).To(BeFalse())
```

File: controllers/release/adapter_test.go [hunk 9]
```diff
@@ -1275,27 +1283,73 @@ var _ = Describe("Release adapter", Ordered, func() {
...
 			})
 		})
+
+		It("returns invalid and an error if the Release has the automated label and no author", func() {
+			adapter.release.Labels = map[string]string{
+				metadata.AutomatedLabel: "false",
+				metadata.AuthorLabel:    "",
+			}
+			valid, err := adapter.validateAuthor()
+			Expect(!valid).To(BeTrue())
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:1330]:
```suggestion
			Expect(valid).To(BeFalse())
```

File: controllers/release/adapter_test.go [hunk 10]
```diff
@@ -1304,25 +1358,81 @@ var _ = Describe("Release adapter", Ordered, func() {
...
+		It("should return invalid and no error if any of the resources are not found", func() {
+			adapter.ctx = toolkit.GetMockedContext(ctx, []toolkit.MockData{
+				{
+					ContextKey: loader.ProcessingResourcesContextKey,
+					Err:        errors.NewNotFound(schema.GroupResource{}, ""),
+				},
+			})
+
+			valid, err := adapter.validateProcessingResources()
+			Expect(!valid).To(BeTrue())
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:1392]:
```suggestion
			Expect(valid).To(BeFalse())
```

File: controllers/release/adapter_test.go [hunk 11]
```diff
@@ -1304,25 +1358,81 @@ var _ = Describe("Release adapter", Ordered, func() {
...
+		It("should return invalid and no error if the ReleasePlanAdmission is found to be disabled", func() {
+			adapter.ctx = toolkit.GetMockedContext(ctx, []toolkit.MockData{
+				{
+					ContextKey: loader.ReleasePlanAdmissionContextKey,
+					Err:        fmt.Errorf("auto-release label set to false"),
+				},
+			})
+
+			valid, err := adapter.validateProcessingResources()
+			Expect(!valid).To(BeTrue())
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:1405]:
```suggestion
			Expect(valid).To(BeFalse())
```

File: controllers/release/adapter_test.go [hunk 12]
```diff
@@ -1304,25 +1358,81 @@ var _ = Describe("Release adapter", Ordered, func() {
...
+		It("should return invalid and no error if multiple ReleasePlanAdmissions exist", func() {
+			adapter.ctx = toolkit.GetMockedContext(ctx, []toolkit.MockData{
+				{
+					ContextKey: loader.ReleasePlanAdmissionContextKey,
+					Err:        fmt.Errorf("multiple ReleasePlanAdmissions found"),
+				},
+			})
+
+			valid, err := adapter.validateProcessingResources()
+			Expect(!valid).To(BeTrue())
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:1418]:
```suggestion
			Expect(valid).To(BeFalse())
```

File: controllers/release/adapter_test.go [hunk 13]
```diff
@@ -1304,25 +1358,81 @@ var _ = Describe("Release adapter", Ordered, func() {
...
+					ContextKey: loader.ProcessingResourcesContextKey,
+					Err:        fmt.Errorf("internal error"),
+					Resource: &loader.ProcessingResources{
+						ReleasePlanAdmission: releasePlanAdmission,
+					},
+				},
+			})
+
+			valid, err := adapter.validateProcessingResources()
+			Expect(!valid).To(BeTrue())
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:1434]:
```suggestion
			Expect(valid).To(BeFalse())
```

File: controllers/release/adapter_test.go [hunk 14]
```diff
@@ -1304,25 +1358,81 @@ var _ = Describe("Release adapter", Ordered, func() {
...
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should return invalid and an error if some other type of error occurs", func() {
+			adapter.ctx = toolkit.GetMockedContext(ctx, []toolkit.MockData{
+				{
+					ContextKey: loader.ProcessingResourcesContextKey,
+					Err:        fmt.Errorf("internal error"),
+					Resource: &loader.ProcessingResources{
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:1427]: It seems that is not properly aligned, but maybe it's just in GitHub. I noticed something similar today.

File: controllers/release/adapter_test.go [hunk 15]
```diff
@@ -1275,27 +1283,73 @@ var _ = Describe("Release adapter", Ordered, func() {
 				}
 			})
 
-			It("returns an error if the ReleasePlan has no author", func() {
+			It("returns invalid and an error if automated label is present but is not set in release status", func() {
+				valid, err := adapter.validateAuthor()
+				Expect(!valid).To(BeTrue())
```
Comment (davidmogar) [inline controllers/release/adapter_test.go:1288]:
```suggestion
			Expect(valid).To(BeFalse())
```

**Review by davidmogar — CHANGES_REQUESTED** [review]

File: controllers/release/adapter_test.go (hunk 14 above)
Comment (johnbieren) [reply controllers/release/adapter_test.go:1427]: that is how the ide wants it

File: controllers/release/adapter.go (hunk 4 above)
Comment (johnbieren) [reply controllers/release/adapter.go:265]: fixed

File: controllers/release/adapter_test.go (hunk 5 above)
Comment (johnbieren) [reply controllers/release/adapter_test.go]: fixed

File: controllers/release/adapter_test.go (hunk 8 above)
Comment (johnbieren) [reply controllers/release/adapter_test.go]: fixed

File: controllers/release/adapter_test.go (hunk 9 above)
Comment (johnbieren) [reply controllers/release/adapter_test.go]: fixed

File: controllers/release/adapter_test.go (hunk 10 above)
Comment (johnbieren) [reply controllers/release/adapter_test.go]: fixed

File: controllers/release/adapter_test.go (hunk 11 above)
Comment (johnbieren) [reply controllers/release/adapter_test.go]: fixed

File: controllers/release/adapter_test.go (hunk 12 above)
Comment (johnbieren) [reply controllers/release/adapter_test.go]: fixed

File: controllers/release/adapter_test.go (hunk 15 above)
Comment (johnbieren) [reply controllers/release/adapter_test.go]: fixed

**Review by davidmogar — APPROVED** [review]

File: controllers/release/adapter_test.go (hunk 7 above)
Comment (johnbieren) [reply controllers/release/adapter_test.go:670]: https://issues.redhat.com/browse/HACBS-2580

**Pull request merged**